### ISSUE TITLE
Introduce consensus param to enable PBTS (in genesis, in updates, etc), along with all validation to update it

### ISSUE THREAD

**Issue by sergio-mena** [cometbft/cometbft#2197]

**Comment by cason** [comment]
So, general question here: what is the expected behaviour for this parameter, meaning which value to use when:

1. A chain starts from genesis, PBTS is enabled from scratch
2. A chain starts from genesis, PBTS is disabled (possibly forever)
3. A chain starts from genesis, PBTS should be enabled from a given height `H`
4. An existing chain (i.e., behaviour when default values is adopted)
5. An existing chain that wants to adopt PBTS immediately after the upgrade
6. An existing chain that wants to adopt PBTS from a given height `H`
7. An existing chain that does not want to adopt PBTS (possibly forever)

**Comment by sergio-mena** [comment]
@cason take a look at all the test cases inside [`TestConsensusParamsUpdate_VoteExtensionsEnableHeight`](https://github.com/cometbft/cometbft/blob/feature/pbts/types/params_test.go#L375).
Let me know if any of your questions above is not answered by that unit test.

**Comment by cason** [comment]
So, correct me if I am wrong:

1. A chain starts from genesis, PBTS is enabled from scratch:`enable_height = 1` (is this the default?)
2. A chain starts from genesis, PBTS is disabled (possibly forever): `enable_height = 0` (is this the default?)
3. A chain starts from genesis, PBTS should be enabled from a given height `H`: `enable_height = H`, provided `H >` current height
4. An existing chain (i.e., behaviour when default values is adopted): if `enable_height = 0` is the default, the existing chain does not (immediately) adopts PBTS
5. An existing chain that wants to adopt PBTS immediately after the upgrade: `enable_height = H+1`, where `H` is the upgrade height?
6. An existing chain that wants to adopt PBTS from a given height `H`: `enable_height = H`, provided `H >` current height
7. An existing chain that does not want to adopt PBTS (possibly forever): `enable_height = 0` (is this the default?)

I think the pending to define here is the default value for existing chains and chains starting from genesis.

**Comment by cason** [comment]
> @cason take a look at all the test cases inside [`TestConsensusParamsUpdate_VoteExtensionsEnableHeight`](https://github.com/cometbft/cometbft/blob/feature/pbts/types/params_test.go#L375).
> Let me know if any of your questions above is not answered by that unit test.

Yes, this defines the rules for upgrading  the `enable_height` parameter. It does not define the behaviour.

**Comment by lasarojc** [comment]
> 🟥   A chain starts from genesis, PBTS is enabled from scratch:enable_height = 1 (is this the default?)
> 🟩  A chain starts from genesis, PBTS is disabled (possibly forever): enable_height = 0 (is this the default?)

The default right now is 0.

> 🟩  A chain starts from genesis, PBTS should be enabled from a given height H: enable_height = H, provided H > current height

It will be the case, since current height is 0. Otherwise we fall into the following case.

> 🟩  An existing chain (i.e., behaviour when default values is adopted): if enable_height = 0 is the default, the existing chain does not (immediately) adopts PBTS
> 🟨  An existing chain that wants to adopt PBTS immediately after the upgrade: enable_height = H+1, where H is the upgrade height?

If the chain is already running v1 and we are talking about a different upgrade that is independent from PBTS, then on.
If the upgrade is to v1, then the upgrade must happen first so the parameter is available to be set.

> 🟩 An existing chain that wants to adopt PBTS from a given height H: enable_height = H, provided H > current height

Same as the 4th case

> 🟩 An existing chain that does not want to adopt PBTS (possibly forever): enable_height = 0 (is this the default?)

Yes

**Comment by sergio-mena** [comment]
> It does not define the behaviour.

The behavior is defined easily:
* if enable_height is currently 0 --> BFT Time
* if enable_height is currently H and H is bigger than the current height --> BFT Time
* otherwise PBTS

**Comment by cason** [comment]
I think this is done, isn't it?

**Comment by lasarojc** [comment]
Closing. Pls reopen if it was hasty.
